pkg/volume/scaleio/sio_util.go

```diff
@@ -0,0 +1,239 @@
+		volumeName:       "volumeName",
+		fsType:           "fsType",
+		readOnly:         "readOnly",
+		username:         "username",
+		password:         "password",
+	}
+
+	secretNotFoundErr       = errors.New("secret not found")
+	configMapNotFoundErr    = errors.New("configMap not found")
+	gatewayNotProvidedErr   = errors.New("gateway not provided")
+	secretRefNotProvidedErr = errors.New("secret ref not provided")
+	systemNotProvidedErr    = errors.New("secret not provided")
+)
+
+// mapScaleIOVolumeSource maps attributes from a ScaleIOVolumeSource to config
+func mapVolumeSource(config map[string]string, source *api.ScaleIOVolumeSource) {
+	config[confKey.gateway] = source.Gateway
+	config[confKey.secretRef] = func() string {
+		if source.SecretRef != nil {
+			return string(source.SecretRef.Name)
+		}
+		return ""
+	}()
+	config[confKey.system] = source.System
+	config[confKey.volumeName] = source.VolumeName
+	config[confKey.sslEnabled] = strconv.FormatBool(source.SSLEnabled)
+	config[confKey.protectionDomain] = source.ProtectionDomain
+	config[confKey.storagePool] = source.StoragePool
+	config[confKey.storageMode] = source.StorageMode
+	config[confKey.sdcRootPath] = source.SDCRootPath
+	config[confKey.fsType] = source.FSType
+	config[confKey.readOnly] = strconv.FormatBool(source.ReadOnly)
+
+	//optionals
+	applyConfigDefaults(config)
+}
+
+func validateConfigs(config map[string]string) error {
+	if config[confKey.gateway] == "" {
+		return gatewayNotProvidedErr
+	}
+	if config[confKey.secretRef] == "" {
+		return secretRefNotProvidedErr
+	}
+	if config[confKey.system] == "" {
+		return systemNotProvidedErr
+	}
+
+	return nil
+}
+
+//TODO remove, default check is done by k8s type checks
+func applyConfigDefaults(config map[string]string) {
+	b, _ := strconv.ParseBool(config[confKey.sslEnabled])
+	config[confKey.sslEnabled] = strconv.FormatBool(b)
+	config[confKey.protectionDomain] = defaultString(config[confKey.protectionDomain], "default")
+	config[confKey.storagePool] = defaultString(config[confKey.storagePool], "default")
+	config[confKey.storageMode] = defaultString(config[confKey.storageMode], "ThinProvisioned")
+	config[confKey.sdcRootPath] = defaultString(config[confKey.sdcRootPath], "/opt/emc/scaleio/sdc/bin/")
+	config[confKey.fsType] = defaultString(config[confKey.fsType], "xfs")
+	b, _ = strconv.ParseBool(config[confKey.readOnly])
+	config[confKey.readOnly] = strconv.FormatBool(b)
+}
+
+func defaultString(val, defVal string) string {
+	if val == "" {
+		return defVal
+	}
+	return val
+}
+
+// saveConfigMap creates/stores a configMap from the config data
+func saveConfigMap(plug *sioPlugin, data map[string]string) error {
+	configName := data[confKey.volumeName]
+	glog.V(4).Info(log("saving confing data as %v configMap", configName))
+
+	kubeClient := plug.host.GetKubeClient()
+	var configMap *api.ConfigMap
+	configMap, err := kubeClient.Core().ConfigMaps(sioDefaultNamespace).Get(configName, meta.GetOptions{})
```

Comment:
use default namespace is bad if you have no privilege to update it or cause collision.